### ISSUE TITLE
Unusual bitflag-value and unreachable code in derwalk()

### ISSUE THREAD

**Issue by adriaandegroot** [vanrein/quick-der#20]
In lib/der_unpack.c, function derwalk(), there's the following code:
```
				if (*walk & DER_PACK_STORE) {
					memset (outarray + (*outctr)++,
```

The DER_PACK_STORE is a weird "bit flag" since it has value 0x00 (see api.h), Possibly this is intended:
((*walk & DER_PACK_MATCHBITS) == DER_PACK_STORE), but it could just as easily be intended the other way around -- I can't quickly tell why that memset is there.

**Comment by vanrein** [comment]
Thanks again; you caught *the other* but in the code.  Ahem.

It's an outlandish bit of code for sure; dealing with empty data at the end of a DER blob.  In case that data is requested to be stored, it should accept that no data is provided, and fill in zero bytes for it.

One of the bits in the command can be either `DER_PACK_ENTER` or `DER_PACK_STORE` to indicate either entering a structure and continue to parse its contents, or store the current structure and skip over it.

The right formulation to do this would have been
```
if ((*walk & ~DER_PACK_MATCHBITS) == DER_PACK_STORE) {
    memset(...);
    smile();
}
```

I was trying to be clear by being verbose, but went off track here.  Thanks for pointing out the middle of the road!

Good bug reports, by the way... I had no way of talking out of them :-D

**Comment by vanrein** [comment]
Fixed in 067ca2bb00e95086bfd1e335c1d2e7b4283dd301.
